### PR TITLE
Make robust auto-reconnect

### PR DESCRIPTION
**Pull request by dilame**
I have made one significant change - `.connect()` never throws anymore. All the connection errors now emitted as `error` event. It unifies error handling workflow.

### CODE REVIEW

**Comment by dilame** [comment]
I'm not even sure the `error` event makes any sense now. It always emitted along with `disconnect` event, and `disconnect` event always contains an `error` object as `reason` when emitted along with the `error`

File: src/mqtt.base-client.ts [hunk 1]
```diff
@@ -117,29 +139,14 @@ export class MqttBaseClient<
         this.next(StateId.Fatal);
     }
 
-    private _connectResolver?: Resolvable<RegisterClientOptions>;
```
Comment (dilame) [inline src/mqtt.base-client.ts:120]: Have no idea why these changes were added in this commit. I have not made them at all.

File: src/mqtt.client.ts [hunk 2]
```diff
@@ -123,22 +121,39 @@ export class MqttClient<
...
         return this.registerClient(await this.resolveConnectOptions());
     }
 
+    public async connect(options?: Resolvable<RegisterClientOptions>) {
+        try {
+            return await this._connect(options);
+        } catch (e) {
+            this.mqttDebug(`Connection error`, e.message);
+            this.emitError(e);
+            return e;
```
Comment (Nerixyz) [inline src/mqtt.client.ts]: Wouldn't this return a potential error in `transport.connect()`?

File: src/mqtt.client.ts [hunk 3]
```diff
@@ -123,22 +121,39 @@ export class MqttClient<
...
         this.connectResolver = options;
         this.setConnecting();
-
-        await this.transport.connect();
+        try {
+            await this.transport.connect();
+        } catch (e) {
+            this.mqttDebug(`Transport connect error ("${this.transport.constructor.name}")`, e.message);
+            if (this.shouldReconnect()) {
+                return this.setDisconnected(e);
```
Comment (Nerixyz) [inline src/mqtt.client.ts]: Maybe add a comment here on why this is done. As I understand, it will return the new promise from calling `connect()` again, but it's confusing seeing the same call twice.

File: src/mqtt.client.ts (hunk 2 above)
Comment (dilame) [reply src/mqtt.client.ts]: Yes, it will **return** error thrown by `transport.connect()`, but will not throw it. It's by design.

File: src/mqtt.client.ts [hunk 4]
```diff
@@ -443,28 +461,28 @@ export class MqttClient<
...
+        const willReconnect = this.shouldReconnect();
+        this.mqttDebug(`Disconnected. Will reconnect: ${willReconnect}. Reconnect attempt #${this.reconnectAttempt}`);
         this.reconnectAttempt++; // this should range from 1 to maxAttempts + 1 when shouldReconnect() is called
-        const willReconnect = this.autoReconnect && this.shouldReconnect();
-
         this.stopExecutingFlows(new AbortError('Client disconnected.'));
 
-        super.setDisconnected();
-        this.emitDisconnect(`reason: ${reason} willReconnect: ${willReconnect}`);
+        this._setDisconnected();
```
Comment (Nerixyz) [inline src/mqtt.client.ts:489]: Why is this prefixed? It should be fine to call the super-method. Or it would need to be renamed.

File: src/mqtt.client.ts (hunk 2 above)
Comment (Nerixyz) [reply src/mqtt.client.ts]: Then it's probably best to specify the return type explicitly.

File: src/mqtt.client.ts (hunk 3 above)
Comment (dilame) [reply src/mqtt.client.ts]: in the first case we explicitly return `this.setDisconnected(e)`, it is done because `setDisconnected` calls `reconnect` that calls `connect`, so we return `connect` result anyway.
In the second case we just call `setDisconnected(e)`  and throw an error. 
I was thinking about caching setDisconnected result in variable but it would not be that expressive, and it's one extra line of code.
Not sure what comment to add in code in this case, at my glance it's self-descripted. What do you think to add?

File: src/mqtt.client.ts (hunk 4 above)
Comment (dilame) [reply src/mqtt.client.ts:489]: Because of TypeScript error. It complained about the return-type. And there was 2 ways - either change `super. setDisconnected` signature or rename it. Changing signature is not right from the design perspective - parent method should not be changed just because child method needed it. The only right choice, in this case, is to rename.

File: src/mqtt.client.ts (hunk 3 above)
Comment (Nerixyz) [reply src/mqtt.client.ts]: > Not sure what comment to add in code in this case, at my glance it's self-descripted. What do you think to add?

Something like 
```
// as the client should reconnect afterwards, this method will call `reconnect` which calls `connect`
```

File: src/mqtt.client.ts (hunk 2 above)
Comment (dilame) [reply src/mqtt.client.ts]: I'm generally an opponent of the `any` type, I am always in favor of explicitly specifying types. But previous `connect()` method signature is `Promise<any>`, so i don't know the actual return type of this method, i am still not an expert in `MQTTS`. Specifying `Promise<any | Error>`  makes no sense since it is the same as `Promise<any>`
If you could help me to specify `_connect` return type - i will be glad to specify `connect()` return type

File: src/mqtt.client.ts (hunk 4 above)
Comment (dilame) [reply src/mqtt.client.ts:489]: BTW MqttClient actually is not a child of MqttBase from the OOP point of view. `Base` is not an entity at its core. `MqttBase` is more of `MqttConnectionManager` or `MqttConnectionStateManager`, but with a leaked abstraction in the current implementation.  `MqttClient` should [contain it but not being inherited from it](https://en.wikipedia.org/wiki/Composition_over_inheritance).

The parent and child methods signatures incompatibility is a kinda litmus test showing this design doesn't reflect requirements.

File: src/mqtt.client.ts (hunk 2 above)
Comment (Nerixyz) [reply src/mqtt.client.ts]: It's "bubbling" from `getConnectFlow` which specifies `unknown` as the success value. This could be changed to `ConnectResponsePacket` (and thus subsequent uses).

File: src/mqtt.client.ts (hunk 4 above)
Comment (Nerixyz) [reply src/mqtt.client.ts:489]: Oh sure

**Comment by dilame** [comment]
Anything else to improve here?

**Comment by Nerixyz** [comment]
> 
> 
> Anything else to improve here?

Seems fine, I'll look at the tests soon.

**Comment by Nerixyz** [comment]
I still don't really understand why `connect` would resolve if there's an error in the transport.

**Comment by dilame** [comment]
> I still don't really understand why `connect` would resolve if there's an error in the transport.

Because it unifies error handling flow. 

What is the difference between the first connect that was initiated by calling `.connect()` method in userland and the second connect initiated after the connection was lost? No difference. It is the same error.

But how a user can catch the connect error that occurred after the connection was lost on the auto-reconnect? The only way to do it is through some event. But the error shipped with the event will be the same error that would throw on the first `.connect()`. 
So it turns out the user should handle the same scenario to catch the same error in 2 ways, and it is required. Ha can't choose the one way, he has to use both.


```ts
declare const mqtts: MqttsClient;

declare function mqttsConnectionErrorHandler(error: Error); 

try {
  await mqtts.connect()
} catch (e) {
  // initial error handling
  mqttsConnectionErrorHandler(e)
}
// subsequent error handling
mqtts.on('error', e => mqttsConnectionErrorHandler(e))

```

It is awful! So we as maintainers should provide one explicit way to handle errors. We can't refuse event emitter here, so that's why I chose it and refuse to throw error from `.connect()`.

This approach is pretty-same as NodeJS Sockets. You call `connect()` first - it always returns you a socket. Then if something is not ok it emits `error` event.

https://github.com/Nerixyz/mqtts/blob/master/src/transport/websocket.transport.ts#L34

**Comment by dilame** [comment]
i suggest using it like this

```ts
declare const mqtts: MqttsClient;
mqtts.on('error', e => handleError(e))
await mqtts.connect()
```

It is simple, clear, and straightforward. It's is like native NodeJS Sockets (ok, almost).

**Comment by dilame** [comment]
Moreover, all you need to handle errors is a `disconnect` event. You don't need the `error` event.

```ts
declare const mqtts: MqttsClient;
mqtts.on('disconnect', event => {
  if(event.reconnect === false) {
    handle_it_manually(event.reason);
  }
})
await mqtts.connect()
```

This structure is all that needed. It covers 100% of cases.
Throwing from `mqtts.connect()` would significantly complicate this structure.

**Comment by dilame** [comment]
i would event say `mqtts.connect()` should return void. At least users should not rely on the `mqtts.connect()` return value.
`mqtts.connect()` just starts the event-oriented mechanism.

**Comment by Nerixyz** [comment]
> 
> 
> i suggest using it like this
> 
> ```ts
> declare const mqtts: MqttsClient;
> mqtts.on('error', e => handleError(e))
> await mqtts.connect()
> ```
> 
> It is simple, clear, and straightforward. It's is like native NodeJS Sockets (ok, almost).

Sure. Then `connect` should return a void-promise. And the test should be adjusted.
